distribution/lib/Standard/Base/0.0.0-dev/src/Data/XML.enso

```diff
@@ -0,0 +1,385 @@
+
+       > Example
+         Get a nested tag:
+
+             XML_Document.from_string '<foo><baz>hello</baz></foo>' . root_element . at 0
+             # => XML_Document.from_string "<baz>hello</baz>" . root_element
+
+       > Example
+         Get a tag attribute.
+
+             root = XML_Document.from_string '<foo bar="one">hello</foo>' . root_element
+             root.at "@bar"
+             # => "one"
+    at : Text | Integer -> Text | XML_Element | Vector (Text | XML_Element) ! No_Such_Key | Index_Out_Of_Bounds | XML_Error
+    at self key:(Text|Integer) =
+        if_missing = case key of
+            _ : Integer -> Error.throw (Index_Out_Of_Bounds.Error key self.num_children)
+            _ : Text -> Error.throw (No_Such_Key.Error self key)
+        self.get key if_missing
+
+    ## Get elements denoted by the given XPath key.
+
+       Arguments:
+       - key: The XPath string to use to search for elements.
+
+       > Example
+         Get an element by xpath.
+
+             root = XML_Document.from_file test_file . root_element
+             root.at "/class/teacher[1]/firstname"
+             # => [XML_Document.from_string "<firstname>Alice</firstname>" . root_element]
+    get_xpath : Text -> Vector (Text | XML_Element) ! XML_Error
+    get_xpath self key:Text =
+        XML_Error.handle_java_exceptions <|
+            xpath = XPathFactory.newInstance.newXPath
+            only_wanted_nodes (nodelist_to_vector (xpath.evaluate key self.java_element XPathConstants.NODESET))
+
+    ## Gets the child elements of an XML element.
+
+       `children` only returns child elements and child text nodes that are not
+       100% whitespace. Other node types, such as comments, are not included.
+
+       > Example:
+             XML_Document.from_string '<foo><baz>hello</baz></foo>' . root_element . children
+             # => [XML_Document.from_string "<baz>hello</baz>"]
+    children : Vector (XML_Element | Text) ! XML_Error
+    children self = only_wanted_nodes self.all_children
+
+    ## Gets the number children of an xml element.
+
+       `num_children` only counts child elements and child text nodes that are
+       not 100% whitespace. Other node types, such as comments, are not included
+       in the count.
+
+       > Example
+         Get the number of children of an element.
+
+             XML_Document.from_string '<foo> <bar>hello</bar> <bar>hello2</bar>< </foo>' . root_element . num_children
+             # => 2
+    num_children : Integer ! XML_Error
+    num_children self = self.children.length
+
+    ## PRIVATE
+       Get the child element indicated by the index.
+
+       `get_child_element` only returns child elements and child text nodes that
+       are not 100% whitespace. Other node types, such as comments, are not
+       considered.
+    get_child_element : Integer -> Any -> Any | Text | XML_Element ! No_Such_Key
+    get_child_element self n:Integer ~if_missing =
+        self.children.get n if_missing
```

Comment:
If I understand the code correctly, this has O(N) complexity. I think users would usually expect a method like this to be O(1).

I understand the issue mainly stems from the fact that we want to exclude the empty children from the index counting, thus we need to know how many of them to skip.

---

I guess it may be fine for now, but for parsing large nodes it may become a bottleneck - an operation iterating N elements using `at` will end up N^2 (because we iterate N elements and each access actually costs us N).

I'd suggest to keep it for now for first iteration but create a ticket to improve it.

---

One way we could improve it is keep a cache for each index, how many empty nodes should be skipped.

Or we can just upon first access cache the index mapping OR just even cache the `children` Vector after first creating it - that can be done using `~lazy` fields.

Reply:
I'll create a ticket.

Reply:
https://github.com/enso-org/enso/issues/7950